### ISSUE TITLE
[3.6.0] splice-subquery optimizer causes query to run 15x slower

### ISSUE THREAD

**Issue by ysmu** [arangodb/arangodb#10853]
## My Environment

* __ArangoDB Version__:        3.6.0
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:   Manual Start in Docker
* __Configuration__:               Nothing special
* __Infrastructure__:               own
* __Operating System__:        Windows 10
* __Total RAM in your machine__:        16Gb
* __Disks in use__:        SSD
* __Used Package__:              Docker - official Docker library

## Steps to reproduce

```py
#!/usr/bin/env python3
import arango


ENDPOINT = 'http://localhost:8529'
DATABASE = 'test'
COLLECTION = 'test'

client = arango.ArangoClient(ENDPOINT)
db = client.db(DATABASE)

try:
    collection = db.create_collection(COLLECTION)
except arango.exceptions.CollectionCreateError as ex:
    # It's ok if the collection is already created.
    if ex.error_code != 1207:
        raise
else:
    # Add initial data.
    collection.add_persistent_index(fields=['a'])
    documents = [{'a': i // 10, 'b': i} for i in range(100_000)]
    collection.import_bulk(documents)


aql = """
    LET list = (
        FOR doc IN test
            COLLECT key = doc.a INTO group = {
                "_key": doc._key,
                "b": doc.b
            }
            RETURN [key, group]
    )
    LET list2 = (
        FOR doc IN test
            COLLECT key = doc.a INTO group = {
                "_key": doc._key,
                "b": doc.b
            }
            RETURN [key, group]
    )

    RETURN [list, list2]
"""


def print_stats(cursor):
    stats = cursor.statistics()
    print('Peak Memory: {}, Execution Time: {}s'.format(stats['peakMemoryUsage'], stats['execution_time']))


print('Running with splice-subqueries optimizer')
print_stats(db.aql.execute(aql))

print()

print('Running without splice-subqueries optimizer')
print_stats(db.aql.execute(aql, optimizer_rules=['-splice-subqueries']))

```

Output
```
Running with splice-subqueries optimizer
Peak Memory: 5215364948, Execution Time: 11.32441782951355s

Running without splice-subqueries optimizer
Peak Memory: 11671035, Execution Time: 0.6482317447662354s
```

**Comment by ysmu** [comment]
Profile (with splice-subqueries)
```
Execution plan:
 Id   NodeType            Calls    Items   Runtime [s]   Comment
  1   SingletonNode           1        1       0.00000   * ROOT
 26   SubqueryStartNode       1        2       0.00000     - LET list = ( /* subquery begin */
 22   IndexNode             101   100001       0.13020       - FOR doc IN test   /* persistent index scan */    
  4   CalculationNode       101   100001       0.02040         - LET #12 = doc.`a`   /* attribute expression */   /* collections used: doc : test */
  5   CalculationNode       101   100001       0.12263         - LET #14 = { "_key" : doc.`_key`, "b" : doc.`b` }   /* simple expression */   /* collections used: doc : test */
  6   CollectNode            11    10001       0.06009         - COLLECT key = #12 INTO group = #14   /* sorted */
  7   CalculationNode        11    10001       0.00879         - LET #16 = [ key, group ]   /* simple expression */
 27   SubqueryEndNode         1        1       0.01314     - RETURN  #16 ) /* subquery end */
 24   SubqueryStartNode       1        2       0.00076     - LET list2 = ( /* subquery begin */
 23   IndexNode             101   100001       0.25558       - FOR doc IN test   /* persistent index scan */    
 12   CalculationNode       101   100001       0.02316         - LET #18 = doc.`a`   /* attribute expression */   /* collections used: doc : test */
 13   CalculationNode       101   100001       0.13305         - LET #20 = { "_key" : doc.`_key`, "b" : doc.`b` }   /* simple expression */   /* collections used: doc : test */
 14   CollectNode            11    10001       9.62158         - COLLECT key = #18 INTO group = #20   /* sorted */
 15   CalculationNode        11    10001       0.01447         - LET #22 = [ key, group ]   /* simple expression */
 25   SubqueryEndNode         1        1       2.41532     - RETURN  #22 ) /* subquery end */
 18   CalculationNode         1        1       0.00493     - LET #24 = [ list, list2 ]   /* simple expression */
 19   ReturnNode              1        1       0.00000     - RETURN #24

Indexes used:
 By   Name                      Type         Collection   Unique   Sparse   Selectivity   Fields    Ranges
 22   idx_1655565997367623680   persistent   test         false    false        10.00 %   [ `a` ]   *
 23   idx_1655565997367623680   persistent   test         false    false        10.00 %   [ `a` ]   *

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   move-calculations-up-2
  3   use-indexes
  4   use-index-for-sort
  5   splice-subqueries

Query Statistics:
 Writes Exec   Writes Ign   Scan Full   Scan Index   Filtered   Exec Time [s]
           0            0           0       200000          0        12.82672

Query Profile:
 Query Stage           Duration [s]
 initializing               0.00000
 parsing                    0.00008
 optimizing ast             0.00001
 loading collections        0.00001
 instantiating plan         0.00006
 optimizing plan            0.00033
 executing                 12.82414
 finalizing                 0.00207
```

**Comment by mchacki** [comment]
Bug confirmed.
Thank you very much for the script to reproduce it.
I am on a fix right now.

**Comment by ysmu** [comment]
I can confirm that the issue is resolved in 3.7. Thanks for fixing the bug! I'm looking forward to trying out some of the new features in 3.7.
